src/main/java/software/amazon/cloudwatchlogs/emf/sinks/AgentSink.java

```diff
@@ -49,10 +62,79 @@ public void accept(MetricsContext context) {
 
         try {
             for (String event : context.serialize()) {
-                client.sendMessage(event + "\n");
+                if (!queue.offer(event)) {
+                    log.warn("Failed to enqueue message into the async buffer. Message was dropped. Ensure the sink destination is available or increase the queue depth to support a larger buffer.");
+                }
             }
         } catch (JsonProcessingException e) {
             log.error("Failed to serialize the metrics with the exception: ", e);
         }
     }
+
+    @Override
+    public CompletableFuture<Void> shutdown() {
+        shutdownRequested.set(true);
+        return shutdownFuture;
+    }
+
+    private CompletableFuture<Void> runSenderThread() {
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        executor.submit(() -> {
+            while (!shutdownRequested.get() || !queue.isEmpty()) {
+                try {
+                    String event = queue.poll(100, TimeUnit.MILLISECONDS);
+                    sendMessageForMaxAttempts(event);
+                } catch (InterruptedException e) {
+                    if (shutdownRequested.get()) {
+                        log.warn("Thread was interrupted and shutdown has been requested.");
+                    }
+                    else {
+                        log.warn("Thread was interrupted but shutdown was not requested.");
+                    }
+                }
+            }
+            completableFuture.complete(null);
+        });
+        return completableFuture;
+    }
+
+    private void sendMessageForMaxAttempts(String message) throws InterruptedException {
+        BoundedFibonacciSequence backoff = null;
+        for (int i = 0; i < Constants.MAX_ATTEMPTS_PER_MESSAGE; i++) {
```

Comment:
The current implementation will never hit the max Fibonacci backoff and will result in a short number of quick socket retries.